### ISSUE TITLE
SPARTA compilation with Kokkos

### ISSUE THREAD

**Issue by ebolandrtx** [sparta/sparta#349]
## Summary

SPARTA Compilation with Kokkos

## Type of Issue

Help Request?

## SPARTA Version (Bug Report)

I am working on compiling SPARTA with the kokkos backend provided in the January 7, 2022 release. When using cmake to build kokkos in the lib/kokkos directory, i've noted that libkokkoscontainers.so and libkokkoscore.so are built, but not libkokkos.so as seems to be expected by SPARTA. Is there anything special i have to include while building Kokkos?

## Expected Behavior (Bug Report)

No error that lkokkos is not found on compilation.

Thanks in advance!

**Comment by stanmoore1** [comment]
@ebolandrtx can you provide more info, like your CMake command? Here is what I see when I compile SPARTA with KOKKOS package enabled:

```
cmake -D PKG_KOKKOS=on ../cmake
make -j4

find . -name "*.a"
./src/KOKKOS/lib/kokkos/core/src/libkokkoscore.a
./src/KOKKOS/lib/kokkos/containers/src/libkokkoscontainers.a
./src/KOKKOS/libpkg_kokkos.a
./src/libsparta_.a
./src/libsparta.a

**Comment by ebolandrtx** [comment]
cmake -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR \
            -DCMAKE_C_COMPILER=gcc \
          -DCMAKE_CXX_COMPILER=$INSTALL_DIR/bin/nvcc_wrapper \
         -DCMAKE_CXX_FLAGS="-fPIC -O2" \
          -DCMAKE_SHARED_LINKER_FLAGS="-fPIC -shared" \
          -DCMAKE_EXE_LINKER_FLAGS="-g -O" \
          -DSPARTA_DEFAULT_CXX_COMPILE_FLAGS="-DSPARTA_PNG -lpng -DSPARTA_GZIP -DSPARTA_FFMPEG" \
          -DBUILD_SHARED_LIBS=ON \
          -DBUILD_PNG=ON \
          -DBUILD_MPI=OFF \
          -DPGK_KOKKOS=ON \
          -DSPARTA_MACHINE="cuda_11.4" \
        $SPARTA_DIR/cmake

**Comment by ebolandrtx** [comment]
Do i need to compile Kokkos first?

**Comment by zhangchonglin** [comment]
I think you just need to enable `kokkos` package using `-DPKG_KOKKOS=ON`, like what Stan mentioned in his comment.

Just in case it will be helpful, I have the following script to build SPARTA with `kokkos` support and it works fine:
https://github.com/zhangchonglin/sparta_scripts/blob/main/build_sparta_kokkos.sh.
You need to modify `-DKokkos_ARCH_PASCAL61=ON` to your GPU device architecture, referring to this: https://github.com/kokkos/kokkos/blob/1fb0c284d458c75370094921d9f202c287502325/BUILD.md#architecture-keywords.

**Comment by ebolandrtx** [comment]
OK - took out a few more of the build directives and i am definitely beyond the build issues. Thanks both of you for responding. Is there an Enable Tests option for a Kokkos based Sparta?

**Comment by stanmoore1** [comment]
>>Is there an Enable Tests option for a Kokkos based Sparta?

Not yet. We do have internal regression tests that run on GPUs though.

**Comment by stanmoore1** [comment]
@ebolandrtx which were the problematic build directives? Shared libs?

**Comment by ebolandrtx** [comment]
@stanmoore1 i believe so. It doesn't seem to be completely set up to build kokkos as a shared library. I also removed -DBuild_Kokkos and just left the -PKG_Kokkos=On in the directives. Not sure that latter really mattered though.

**Comment by ebolandrtx** [comment]
Ok i put my finger on it - was not located in the build directory, so it looked like problems from the SHARED libs. Can SPARTA use multiple GPUs on the same node without openMPI?

**Comment by stanmoore1** [comment]
>> Can SPARTA use multiple GPUs on the same node without openMPI?

No, you must have at least 1 MPI rank per GPU with Kokkos. You can have *multiple* MPI ranks per GPU with CUDA MPS but not multiple GPUs per MPI rank.

**Comment by ebolandrtx** [comment]
@stanmoore1 Thanks! I didn't think fftw supported CUDA. Would you recommend using the native FFT provided with Sparta? Is that FFT MPI capable?

**Comment by stanmoore1** [comment]
The FFTs in SPARTA do not yet have Kokkos support--we are working on that, and this enhancement should be released soon. So currently the FFTs will just run on the host. Best for CUDA is to use NVIDIA's cuFFT library, but again we need to add support for that--in progress.

**Comment by stanmoore1** [comment]
@ebolandrtx Kokkos support for FFTs was added with #451, e.g. you can now use NVIDIA's `cuFFT` library on GPUs.

**Comment by ebolandrtx** [comment]
@stanmoore1 Thanks for responding on this thread. Very cool!

**Comment by stanmoore1** [comment]
@ebolandrtx the shared lib issue with KOKKOS package should be fixed by https://github.com/sparta/sparta/pull/627.
